Package bibutils fails to build on macOS 11.3

Building the package `bibutils` produces a large number of compiler warnings, and then fails with

    ...
    risout.c:111:26: note: previous initialization is here
            [ 0 ... NUM_TYPES-1 ] = 0,
                                    ^
    8 warnings generated.
    ld: unknown option: -soname
    clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
    make[1]: *** [Makefile:83: libbibutils.so] Error 1
    make[1]: Target 'all' not remade because of errors.
    make: *** [Makefile:40: all] Error 2
    builder for '/nix/store/bqwyr2gywrffh3xrwplhi4sv7962d21r-bibutils-6.10.drv' failed with exit code 2
    error: build of '/nix/store/bqwyr2gywrffh3xrwplhi4sv7962d21r-bibutils-6.10.drv' failed

I have built bibutils by hand on macOS, without difficulty (I last tried it on 10.15), so it's not something intrinsically broken in the upstream package.  I hope, therefore, it just requires a minor tweak in the Nix package.

I note from [the package search](https://search.nixos.org/packages?channel=20.09&show=bibutils&from=0&size=50&sort=relevance&query=bibutils) that macOS isn't one of the listed supported platforms.

I'm a little uncertain how best to report my current environment, but:

  * OS : macOS 11.3.1
  * compiler: Apple clang version 11.0.3 (clang-1103.0.32.29) ; Target: x86_64-apple-darwin20.4.0
  * nix --version: nix (Nix) 2.3.9

--- a/pkgs/tools/misc/bibutils/default.nix
+++ b/pkgs/tools/misc/bibutils/default.nix
@@ -1,4 +1,8 @@
-{ lib, stdenv, fetchurl }:
+{ lib
+, stdenv
+, fetchurl
+, static ? stdenv.hostPlatform.isStatic
+}:
 
 stdenv.mkDerivation rec {
   pname = "bibutils";
@@ -9,7 +13,11 @@ stdenv.mkDerivation rec {
     sha256 = "15p4av74ihsg03j854dkdqihpspwnp58p9g1lhx48w8kz91c0ml6";
   };
 
-  configureFlags = [ "--dynamic" "--install-dir" "$(out)/bin" "--install-lib" "$(out)/lib" ];
+  configureFlags = [
+    (if static then "--static" else "--dynamic")
+    "--install-dir" "$(out)/bin"
+    "--install-lib" "$(out)/lib"
+  ];
   dontAddPrefix = true;
 
   doCheck = true;

--- a/pkgs/tools/misc/bibutils/default.nix
+++ b/pkgs/tools/misc/bibutils/default.nix
@@ -13,6 +13,11 @@ stdenv.mkDerivation rec {
     sha256 = "15p4av74ihsg03j854dkdqihpspwnp58p9g1lhx48w8kz91c0ml6";
   };
 
+  preConfigure = lib.optionalString stdenv.isDarwin ''
+    substituteInPlace lib/Makefile.dynamic \
+      --replace '-Wl,-soname,$(SONAME)' ""
+  '';
+
   configureFlags = [
     (if static then "--static" else "--dynamic")
     "--install-dir" "$(out)/bin"
@@ -22,6 +27,9 @@ stdenv.mkDerivation rec {
 
   doCheck = true;
   checkTarget = "test";
+  preCheck = lib.optionalString stdenv.isDarwin ''
+    export DYLD_LIBRARY_PATH=`pwd`/lib
+  '';
 
   meta = with lib; {
     description = "Bibliography format interconversion";
@@ -29,6 +37,6 @@ stdenv.mkDerivation rec {
     homepage = "https://sourceforge.net/p/bibutils/home/Bibutils/";
     license = licenses.gpl2;
     maintainers = [ maintainers.garrison ];
-    platforms = platforms.linux;
+    platforms = platforms.unix;
   };
 }

